File: FromSlides/C/05_loop_decompose_private.c
#include <stdio.h>
#include <omp.h>

#define MAX_ITS 10000

int main()
{
  int nproc, i, sum, thread_id;
  nproc = omp_get_max_threads();
  int its_per_proc[nproc];

  for (i = 0; i< nproc; ++i){
    its_per_proc[i] = 0;
  }

#pragma omp parallel for private(thread_id)
  for (i = 0; i< MAX_ITS; ++i){
    thread_id = omp_get_thread_num();
    its_per_proc[thread_id]++;
}

  sum = 0;
  for (i = 0; i< nproc; ++i){
    printf("Processor %i performed %i iterations\n", i, its_per_proc[i]);
    sum += its_per_proc[i];
  }
  printf("Total work on all processors is %i\n", sum); 
}

File: FromSlides/Fortran/05_loop_decompose_private.f90
PROGRAM loop_decompose
  
  USE omp_lib
  IMPLICIT NONE
  INTEGER, PARAMETER :: max_its = 10000
  INTEGER :: nproc, i, thread_id
  INTEGER, DIMENSION(:), ALLOCATABLE :: its_per_proc

  nproc = omp_get_max_threads()
  ALLOCATE(its_per_proc(nproc))
  its_per_proc = 0

!This causes a SEPARATE copy of the variable thread_id
!to be created on each thread
!$OMP PARALLEL DO PRIVATE(thread_id)
  DO i = 1, max_its
    thread_id = omp_get_thread_num() + 1
    its_per_proc(thread_id) = its_per_proc(thread_id) + 1
  END DO
!$OMP END PARALLEL DO

  DO i = 1, nproc
    PRINT '(A, I0, A, I0, A)', 'Processor ', i, ' performed ', &
        its_per_proc(i), ' iterations'
  END DO

 PRINT '(A, I0)','Total work on all processors is ', SUM(its_per_proc)

 DEALLOCATE(its_per_proc)

END PROGRAM loop_decompose
